tensorflow/go/graph.go

```diff
@@ -147,6 +147,73 @@ func (g *Graph) Operations() []Operation {
 	return ops
 }
 
+// AddGradients adds operations to compute the partial derivatives of sum of `y`s w.r.t `x`s,
+// i.e., d(y_1 + y_2 + ...)/dx_1, d(y_1 + y_2 + ...)/dx_2...
+// This is a simplified version of AddGradientsWithPrefix() without prefix
+func (g *Graph) AddGradients(y []Output, x []Output, dx []Output) ([]Output, error) {
+	return g.AddGradientsWithPrefix("", y, x, dx)
+}
+
+// AddGradientsWithPrefix adds operations to compute the partial derivatives of sum of `y`s w.r.t `x`s,
+// i.e., d(y_1 + y_2 + ...)/dx_1, d(y_1 + y_2 + ...)/dx_2...
+// This is a variant of AddGradients that allows to caller to pass a custom
+// name prefix to the operations added to a graph to compute the gradients.
+func (g *Graph) AddGradientsWithPrefix(prefix string, y []Output, x []Output, dx []Output) ([]Output, error) {
```

Comment:
Let's have just one method: `AddGradients` in Go (which can take a prefix string), instead of having multiple methods.

And then we can add a method to the `op` package, so that `Gradients` appears just like any other op (and can pick up the appropriate prefix from the `Scope`). A similar thing was done for the Java API.

That would allow common usage like:

```go
s := op.NewScope()
x := op.Placeholder(s, tf.Float) // Matrix to be provided to Session.Run
y := op.Mul(s, input, input)
dy := op.Gradients(s, []tf.Output{y}, []tf.Output{x})[0]
```

Reply:
Thanks for reviewing!

Some questions:
1. If the "appropriate prefix from the `Scope`” is `namespace` of `Scope`?
2. In the example, the method name is `Gradients`, is this a suggest name? Or I should still use `AddGradients`?
3. Also in the example, the `Gradients` Method only have three parameters(`scope`, `y[]`, `x[]`), don't have the `dx[]`, is this a suggestion that in the method the `dx[]` should just be set to `nil`?
4. I see all `op` method is in `wrappers.go`, same for this method?

An implementation example:
```
func Gradients(scope *Scope, y []tf.Output, x []tf.Output) (output [] tf.Output) {
	var err error
	if output, err = scope.graph.AddGradients(scope.namespace, y, x, nil); err != nil {
		scope.UpdateErr("Gradients", err)
		return
	}
	return output
}
```

Reply:
Thanks for the PR :)

1. Yup, see https://github.com/tensorflow/tensorflow/blob/7d3884bb87dc02c4548f55749f3d6db1b8364ddc/tensorflow/go/op/scope.go#L75

2. Let's use `Gradients`, as it's akin to a composite TensorFlow operation.

3. Perhaps we can treat `dx` as an optional parameter using the pattern we use for other operations. For example, see [`BatchMatMul`](https://godoc.org/github.com/tensorflow/tensorflow/tensorflow/go/op#BatchMatMul)

4. Nope, `wrappers.go` is automatically generated, I'd just add that to a new file `gradients.go` in the `op` directory.

Sound reasonable?